Optimize Snowflake load_file using native COPY INTO

tatiana opened the pull request:
Fix: #430 

Reduced the time to load to Snowflake by 20% for 5GB datasets (from 24.46 min to 5.49 min). Further details in the PR results file.

Co-authored-by: Ankit Chaurasia <ankit.chaurasia@astronomer.io>

sunank200 commented on src/astro/databases/snowflake.py:
```diff
@@ -291,6 +292,52 @@ def drop_stage(self, stage: SnowflakeStage) -> None:
+        :param source_file: File from which we need to transfer data
+        :param target_table: Table to which the content of the file will be loaded to
+        :param if_exists: Strategy used to load (currently supported: "append" or "replace")
+        :param native_support_kwargs: kwargs to be used by method involved in native support flow
```
Would be it useful for users to know what are different `native_support_kwargs` supported? or may be reference to `COPY INTO` documentation?

tatiana replied:
@sunank200, please, have a look now!
I had already added the reference to the official docs on the bottom of the docstring, do you think I should add some other? Or place it in another place?

sunank200 reviewed (approved):
Looks good to me with a minor comment.

sunank200 commented on tests/databases/test_snowflake.py:
```diff
@@ -450,3 +450,40 @@ def test_create_stage_amazon_fails_due_to_no_credentials(get_credentials):
+    """Load a file to a Snowflake table using the native optimisation."""
+    filepath = remote_files_fixture[0]
+    database, target_table = database_table_fixture
+    database.load_file_to_table(
```
This test should have called `load_file_to_table_natively()`.

sunank200 replied:
Makes sense

sunank200 reviewed (changes requested):
The test is calling wrong method. Thats the reason coverage shows this test not being run.

tatiana commented:
We added the benchmark results to the PR and fixed the broken tests.

Two checks are failing probably because of https://github.com/pypa/pip/pull/11298 - which affects all branches. I just re-run the tests for a branch which was passing 4h ago, that wasn't changed and the same tests failed.

I've contacted TP to get his advice, and in the meantime, we'll merge this PR.

tatiana commented on example_dags/example_amazon_s3_snowflake_transform.py:
```diff
@@ -21,17 +21,17 @@ def combine_data(center_1: Table, center_2: Table):
 
 
-@aql.dataframe(identifiers_as_lower=False)
+@aql.dataframe()
```
We should be able to handle capitalization more consistently across dataframes / loading file after implementing:
https://github.com/astronomer/astro-sdk/issues/564

tatiana commented on example_dags/example_amazon_s3_snowflake_transform.py:
```diff
@@ -67,11 +67,11 @@ def example_amazon_s3_snowflake_transform():
 
     temp_table_1 = aql.load_file(
-        input_file=File(path=f"{s3_bucket}/ADOPTION_CENTER_1.csv"),
+        input_file=File(path=f"{s3_bucket}/ADOPTION_CENTER_1_unquoted.csv"),
```
The CSV was very odd beforehand in the format:
```
"header1","header2"
"value1","value2"
```

The COPY INTO command wasn't acceptable in this format, and we decided that this file could be cleaned beforehand. Users may face similar issues in future.

tatiana commented on tests/benchmark/results.md:
```diff
@@ -95,6 +94,25 @@ The benchmark was run as a Kubernetes job in GKE:
+| snowflake  | hundred_kb | 9.19s        | 45.4MB       | 2.55s           | 120.0ms           |
+| snowflake  | ten_mb     | 10.9s        | 47.51MB      | 2.58s           | 160.0ms           |
+| snowflake  | one_gb     | 1.07min      | 47.94MB      | 8.7s            | 5.67s             |
+| snowflake  | five_gb    | 5.49min      | 53.69MB      | 18.76s          | 1.6s              |
```
I'll log a ticket for us to further investigate why we weren't able to run the benchmark with 10 GB for this implementation.

tatiana commented on src/astro/databases/snowflake.py:
```diff
@@ -291,6 +301,93 @@ def drop_stage(self, stage: SnowflakeStage) -> None:
+        # Snowflake doesn't handle well mixed capitalisation of column name chars
+        # we are handling this more gracefully in a separate PR
+        if dataframe is not None:
+            dataframe.columns.str.upper()
```
We should be able to handle capitalization more consistently across dataframes / loading file after implementing: #564
